daffodil-runtime1/src/main/scala/org/apache/daffodil/runtime1/processors/parsers/BlobLengthParsers.scala

```diff
@@ -55,9 +55,9 @@ trait ByteChunkWriter { self: PrimParser =>
         remainingBitsToGet -= bitsToGet
       } else {
         val remainingBits =
-          if (dis.remainingBits.isDefined) {
+          if (dis.remainingLimitedBits.isDefined) {
             val totalBitsRead = nBits - remainingBitsToGet
-            MaybeULong(dis.remainingBits.get + totalBitsRead)
+            MaybeULong(dis.remainingLimitedBits.get + totalBitsRead)
```

Comment:
Please add a comment explaining why out of 26 places calling `remainingBits`, you decided to change this particular place to call `remainingLimitedBits` instead of `remainingBits`.  The `remainingLimitedBits` function returns only the remaining bits until the limit is reached, so you're losing the information that the remaining bits until the end of the input is reached might be even smaller.

Reply:
Yeah, I think this might actually be incorrect? This use of `remainingBits` here is to calculate a parameter for `PENotEnoughBits`. And since this is for an error message, like the other uses of `remainingBits`, I think we do want to take into account EOF, and so should use `remainingBits`.

Which I think means there are no more uses of `remainingLimitedBits`, and maybe it's logic can just be merge into `remainingBits`?

Another thought, nothing should really use `reaminingBits` to make parsing decision (we don't currently, but we could accidentally in the future). Parsers should always use `isDefinedForLength` if they want to know if bits are available, since that takes into account both bit limit and EOF, and will also read and cache bytes from the input stream as needed. The only place it is reasonable to use is in error messages when `isDefinedForLength` returns false, which is how it's used right now (and it's not used in unparse at all). Since it really shouldn't be used, and is only used for not enough bits diagnostics, I wonder if we can just remove it and move the logic into `PENotEnoughBits` available, or at least rename it so it's clear it should only be used for diagnostics when `isDefinedForLength` fails? This would avoid us accidentally using it for actual parsing decisions.

Reply:
The old `remainingBits` function which only considers end-of-limit got renamed to `remainingLimitedBits`. The new function that returns the length to the end-of-limit or end-of-stream now uses the name of `remainingBits`. I originally considered `reminaingLimitedBits` and `remainingStreamBits` but I didn't think that worked well. So this change isn't really a change to functionality but a refactoring to use the new name.

To fix this bug, I initially found the two locations that output the error message and threw a `RuntimeException` if the order of the numbers was backwards. I updated all instances where the error message was called because it seemed like getting the length to end-of-limit or end-of-file was correct when reporting an error message. 

In this case, it is doing more than reporting an error, so I decided to leave the functionality as-is.

Reply:
> In this case, it is doing more than reporting an error, so I decided to leave the functionality as-is.

I think it is reporting an error, it's just using slightly fancier logic to take into account read data since the blob parsers reads data in chunks.

Reply:
I renamed `bytesAvailable` to `knownBytesAvailable` to help indicate that this is not necessarily a known value and should not be relied on in parsing situations. The `remainingBits`/`remainingLimitBits` have been refactored out and that logic is now in the error reporting function.